File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.py
class Solution:
    def minFlips(self, s: str) -> int:
        n = len(s)
        s = s + s
        target = "01" * n
        res = n
        cnt = 0

        for i in range(2 * n):
            if s[i] != target[i]: 
                cnt += 1
            
            if i >= n:
                if s[i - n] != target[i - n]:
                    cnt -= 1
            
            if i >= n - 1:
                res = min(res, cnt, n - cnt)

        return res

File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.cpp
class Solution {
public:
    int minFlips(string s) {
        int n = s.size();
        s += s;
        int res = n;
        int cnt = 0;
        string target;
        target.reserve(2 * n);
        for (int i = 0; i < n; i++) {
            target += "01"; 
        }

        for(int i = 0; i < 2 * n; i++) {
            if(s[i] != target[i]) {
                cnt++;
            }

            if(i >= n) {
                if(s[i - n] != target[i - n]){
                    cnt--;
                }
            }

            if(i >= n - 1) {
                res = min({res, cnt, n - cnt});
            }
        }

        return res;
    }
};